File: spec/wuzzy-crawler/crawl-tasks.spec.lua
local codepath = 'wuzzy-crawler.wuzzy-crawler'

describe('Wuzzy-Crawler Crawl-Tasks', function()
  local WuzzyCrawler = require(codepath)
  before_each(function()
    CacheOriginalGlobals()
    WuzzyCrawler = require(codepath)
  end)
  after_each(function()
    RestoreOriginalGlobals()
    package.loaded[codepath] = nil
    package.loaded['..common.acl'] = nil
  end)

  describe('Add-Crawl-Tasks', function()
    it('rejects Add-Crawl-Tasks messages from unknown callers', function()
      _G.send = spy.new(function() end)

      assert.has_error(function()
        GetHandler('Add-Crawl-Tasks').handle({
          from = 'alice-address',
          data = 'mock-crawl-task'
        })
      end, 'Permission Denied')
    end)

    it('requires Crawl Task Data', function()
      _G.send = spy.new(function() end)

      assert.has_error(function()
        GetHandler('Add-Crawl-Tasks').handle({ from = _G.owner })
      end, 'Missing Crawl Task Data')
    end)

    it('validates Add-Crawl-Tasks Task Data', function()
      _G.send = spy.new(function() end)
      local crawlTaskData = 'invalid-url'

      assert.has_error(function()
        GetHandler('Add-Crawl-Tasks').handle({
          from = _G.owner,
          data = crawlTaskData
        })
      end, 'Invalid Crawl Task Data: ' .. crawlTaskData)
    end)

    it('rejects duplicate crawl tasks', function()
      _G.send = spy.new(function() end)
      local url = 'https://example.com/some/path'

      GetHandler('Add-Crawl-Tasks').handle({
        from = _G.owner,
        data = url
      })

      assert.has_error(function()
        GetHandler('Add-Crawl-Tasks').handle({
          from = _G.owner,
          data = url
        })
      end, 'Duplicate Crawl Task: ' .. url)
    end)

    it('accepts Add-Crawl-Tasks message from owner, admin, acl', function()
      _G.send = spy.new(function() end)
      local aliceAddress = 'alice-address'
      local bobAddress = 'bob-address'
      GetHandler('Update-Roles').handle({
        from = _G.owner,
        data = require('json').encode({
          Grant = {
            [aliceAddress] = { 'admin' },
            [bobAddress] = { 'Add-Crawl-Tasks' }
          }
        })
      })
      _G.send:clear()
      local tasks = {
        'arns://memeticblock',
        'arns://wuzzy',
        'arns://cookbook',
        'arns://cookbook_ao'
      }

      GetHandler('Add-Crawl-Tasks').handle({
        from = _G.owner,
        data = tasks[1] .. '\n' .. tasks[2]
      })
      assert.spy(_G.send).was.called_with({
        target = _G.owner,
        action = 'Add-Crawl-Tasks-Result',
        data = 'OK'
      })

      GetHandler('Add-Crawl-Tasks').handle({
        from = aliceAddress,
        data = tasks[3]
      })
      assert.spy(_G.send).was.called_with({
        target = _G.owner,
        action = 'Add-Crawl-Tasks-Result',
        data = 'OK'
      })

      GetHandler('Add-Crawl-Tasks').handle({
        from = bobAddress,
        data = tasks[4]
      })
      assert.spy(_G.send).was.called_with({
        target = _G.owner,
        action = 'Add-Crawl-Tasks-Result',
        data = 'OK'
      })

      assert.are_same({
        [tasks[1]] = {
          AddedBy = _G.owner,
          SubmittedUrl = tasks[1],
          URL = tasks[1],
          Protocol = 'arns',
          Domain = 'memeticblock',
          Path = ''
        },
        [tasks[2]] = {
          AddedBy = _G.owner,
          SubmittedUrl = tasks[2],
          URL = tasks[2],
          Protocol = 'arns',
          Domain = 'wuzzy',
          Path = ''
        },
        [tasks[3]] = {
          AddedBy = aliceAddress,
          SubmittedUrl = tasks[3],
          URL = tasks[3],
          Protocol = 'arns',
          Domain = 'cookbook',
          Path = ''
        },
        [tasks[4]] = {
          AddedBy = bobAddress,
          SubmittedUrl = tasks[4],
          URL = tasks[4],
          Protocol = 'arns',
          Domain = 'cookbook_ao',
          Path = ''
        }
      }, WuzzyCrawler.State.CrawlTasks)
    end)

    pending('uses ~patch@1.0 whenever updating state')
  end)

  describe('Remove-Crawl-Tasks', function()
    it('rejects Remove-Crawl-Tasks message from unknown callers', function()
      _G.send = spy.new(function() end)

      assert.has_error(function()
        GetHandler('Remove-Crawl-Tasks').handle({
          from = 'alice-address',
          data = 'mock-crawl-task'
        })
      end, 'Permission Denied')
    end)

    it('requires Remove-Crawl-Tasks Task Data', function()
      _G.send = spy.new(function() end)

      assert.has_error(function()
        GetHandler('Remove-Crawl-Tasks').handle({ from = _G.owner })
      end, 'Missing Crawl Task Data to remove')
    end)

    it('validates Remove-Crawl-Tasks Task Data', function()
      local url = 'arns://memeticblock'

      assert.has_error(function()
        GetHandler('Remove-Crawl-Tasks').handle({
          from = _G.owner,
          data = url
        })
      end, 'Crawl Task not found: ' .. url)
    end)

    it('accepts Remove-Crawl-Tasks message from owner, admin, acl', function()
      _G.send = spy.new(function() end)
      local aliceAddress = 'alice-address'
      local bobAddress = 'bob-address'
      GetHandler('Update-Roles').handle({
        from = _G.owner,
        data = require('json').encode({
          Grant = {
            [aliceAddress] = { 'admin' },
            [bobAddress] = { 'Remove-Crawl-Tasks' }
          }
        })
      })
      _G.send:clear()
      local tasks = {
        'arns://memeticblock',
        'arns://wuzzy',
        'arns://cookbook',
        'arns://cookbook_ao'
      }

      GetHandler('Add-Crawl-Tasks').handle({
        from = _G.owner,
        data = tasks[1] .. '\n' ..
          tasks[2] .. '\n' ..
          tasks[3] .. '\n' ..
          tasks[4]
      })
      assert.is_not_nil(WuzzyCrawler.State.CrawlTasks[tasks[1]])
      assert.is_not_nil(WuzzyCrawler.State.CrawlTasks[tasks[2]])
      assert.is_not_nil(WuzzyCrawler.State.CrawlTasks[tasks[3]])
      assert.is_not_nil(WuzzyCrawler.State.CrawlTasks[tasks[4]])

      GetHandler('Remove-Crawl-Tasks').handle({
        from = _G.owner,
        data = tasks[2]
      })
      assert.is_nil(WuzzyCrawler.State.CrawlTasks[tasks[2]])

      GetHandler('Remove-Crawl-Tasks').handle({
        from = aliceAddress,
        data = tasks[3]
      })
      assert.is_nil(WuzzyCrawler.State.CrawlTasks[tasks[3]])

      GetHandler('Remove-Crawl-Tasks').handle({
        from = bobAddress,
        data = tasks[1]
      })
      assert.is_nil(WuzzyCrawler.State.CrawlTasks[tasks[1]])

      assert.are_same({
        [tasks[4]] = {
          AddedBy = _G.owner,
          SubmittedUrl = tasks[4],
          URL = tasks[4],
          Protocol = 'arns',
          Domain = 'cookbook_ao',
          Path = ''
        }
      }, WuzzyCrawler.State.CrawlTasks)
    end)

    pending('uses ~patch@1.0 whenever updating state')
  end)
end)

File: test/spec/contracts/wuzzy-crawler/crawl-tasks.spec.ts
import { expect } from 'chai'

import {
  ALICE_ADDRESS,
  AOTestHandle,
  BOB_ADDRESS,
  createLoader,
  OWNER_ADDRESS
} from '~/test/util/setup'

describe('Wuzzy-Crawler Crawl-Tasks', () => {
  let handle: AOTestHandle

  beforeEach(async () => {
    handle = (await createLoader('wuzzy-crawler')).handle
  })

  describe('Add-Crawl-Tasks', () => {
    it('rejects Add-Crawl-Tasks messages from unknown callers', async () => {
      const result = await handle({
        From: ALICE_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: 'fake crawl task'
      })

      expect(result.Error).to.exist
      expect(result.Error).to.include('Permission Denied')
    })

    it('validates Add-Crawl-Tasks messages', async () => {
      const noDataResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ]
      })
      expect(noDataResult.Error).to.exist
      expect(noDataResult.Error).to.include('Missing Crawl Task Data')

      const invalidUrl = '1234'
      const invalidUrlResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: invalidUrl
      })
      expect(invalidUrlResult.Error).to.exist
      expect(invalidUrlResult.Error).to.include(
        `Invalid Crawl Task Data: ${invalidUrl}`
      )
    })

    it('temporarily rejects Add-Crawl-Tasks for http/https urls', async () => {
      const httpUrl = 'http://example.com'
      const httpsUrl = 'https://example.com'

      const httpResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: httpUrl
      })
      expect(httpResult.Error).to.exist
      expect(httpResult.Error).to.include(
        `The http protocol is not yet supported: ${httpUrl}`
      )

      const httpsResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: httpsUrl
      })
      expect(httpsResult.Error).to.exist
      expect(httpsResult.Error).to.include(
        `The https protocol is not yet supported: ${httpsUrl}`
      )
    })

    it('rejects duplicate crawl tasks', async () => {
      const url = 'arns://memeticblock'
      await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: url
      })
      const duplicateResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: url
      })
      expect(duplicateResult.Error).to.exist
      expect(duplicateResult.Error).to.include(`Duplicate Crawl Task: ${url}`)
    })

    it('accepts Add-Crawl-Tasks messages from Owner, admin, ACL', async () => {
      await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Update-Roles' } ],
        Data: JSON.stringify({
          Grant: {
            [ALICE_ADDRESS]: [ 'admin' ],
            [BOB_ADDRESS]: [ 'Add-Crawl-Tasks' ]
          }
        })
      })

      const ownerResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: 'arns://memeticblock'
      })
      expect(ownerResult.Error).to.not.exist
      expect(ownerResult.Messages).to.have.lengthOf(1)
      expect(ownerResult.Messages[0].Target).to.equal(OWNER_ADDRESS)
      expect(ownerResult.Messages[0].Data).to.equal('OK')
      expect(ownerResult.Messages[0].Tags).to.deep.include({
        name: 'Action',
        value: 'Add-Crawl-Tasks-Result'
      })

      const adminResult = await handle({
        From: ALICE_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: 'arns://wuzzy'
      })
      expect(adminResult.Error).to.not.exist
      expect(adminResult.Messages).to.have.lengthOf(1)
      expect(adminResult.Messages[0].Target).to.equal(ALICE_ADDRESS)
      expect(adminResult.Messages[0].Data).to.equal('OK')
      expect(adminResult.Messages[0].Tags).to.deep.include({
        name: 'Action',
        value: 'Add-Crawl-Tasks-Result'
      })

      const aclResult = await handle({
        From: BOB_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: 'arns://cookbook'
      })
      expect(aclResult.Error).to.not.exist
      expect(aclResult.Messages).to.have.lengthOf(1)
      expect(aclResult.Messages[0].Target).to.equal(BOB_ADDRESS)
      expect(aclResult.Messages[0].Data).to.equal('OK')
      expect(aclResult.Messages[0].Tags).to.deep.include({
        name: 'Action',
        value: 'Add-Crawl-Tasks-Result'
      })

      const viewStateResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'View-State' } ]
      })
      expect(viewStateResult.Error).to.not.exist
      expect(viewStateResult.Messages).to.have.lengthOf(1)
      expect(viewStateResult.Messages[0].Data).to.exist
      const state = JSON.parse(viewStateResult.Messages[0].Data)
      expect(state.CrawlTasks).to.be.an('object')
      expect(Object.keys(state.CrawlTasks)).to.have.lengthOf(3)
      expect(state.CrawlTasks['arns://memeticblock']).to.exist
      expect(state.CrawlTasks['arns://wuzzy']).to.exist
      expect(state.CrawlTasks['arns://cookbook']).to.exist
    })
  })

  describe('Remove-Crawl-Tasks', () => {
    it('rejects Remove-Crawl-Tasks messages from unknown callers', async () => {
      const result = await handle({
        From: ALICE_ADDRESS,
        Tags: [ { name: 'Action', value: 'Remove-Crawl-Tasks' } ],
        Data: 'fake crawl task'
      })

      expect(result.Error).to.exist
      expect(result.Error).to.include('Permission Denied')
    })

    it('validates Remove-Crawl-Tasks messages', async () => {
      const noDataResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Remove-Crawl-Tasks' } ]
      })
      expect(noDataResult.Error).to.exist
      expect(noDataResult.Error).to.include('Missing Crawl Task Data to remove')

      const url = 'arns://memeticblock'
      const nothingToRemoveResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Remove-Crawl-Tasks' } ],
        Data: url
      })
      expect(nothingToRemoveResult.Error).to.exist
      expect(nothingToRemoveResult.Error).to.include(
        `Crawl Task not found: ${url}`
      )
    })

    it('accepts Remove-Crawl-Tasks messages from Owner, admin, ACL', async () => {
      const urls = [ 'arns://memeticblock', 'arns://wuzzy', 'arns://cookbook' ]
      await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Add-Crawl-Tasks' } ],
        Data: urls.join('\n')
      })
      await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Update-Roles' } ],
        Data: JSON.stringify({
          Grant: {
            [ALICE_ADDRESS]: [ 'admin' ],
            [BOB_ADDRESS]: [ 'Remove-Crawl-Tasks' ]
          }
        })
      })

      const ownerResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'Remove-Crawl-Tasks' } ],
        Data: urls[0]
      })
      expect(ownerResult.Error).to.not.exist
      expect(ownerResult.Messages).to.have.lengthOf(1)
      expect(ownerResult.Messages[0].Target).to.equal(OWNER_ADDRESS)
      expect(ownerResult.Messages[0].Data).to.equal('OK')

      const adminResult = await handle({
        From: ALICE_ADDRESS,
        Tags: [ { name: 'Action', value: 'Remove-Crawl-Tasks' } ],
        Data: urls[1]
      })
      expect(adminResult.Error).to.not.exist
      expect(adminResult.Messages).to.have.lengthOf(1)
      expect(adminResult.Messages[0].Target).to.equal(ALICE_ADDRESS)
      expect(adminResult.Messages[0].Data).to.equal('OK')

      const aclResult = await handle({
        From: BOB_ADDRESS,
        Tags: [ { name: 'Action', value: 'Remove-Crawl-Tasks' } ],
        Data: urls[2]
      })
      expect(aclResult.Error).to.not.exist
      expect(aclResult.Messages).to.have.lengthOf(1)
      expect(aclResult.Messages[0].Target).to.equal(BOB_ADDRESS)
      expect(aclResult.Messages[0].Data).to.equal('OK')

      const viewStateResult = await handle({
        From: OWNER_ADDRESS,
        Tags: [ { name: 'Action', value: 'View-State' } ]
      })
      expect(viewStateResult.Error).to.not.exist
      expect(viewStateResult.Messages).to.have.lengthOf(1)
      expect(viewStateResult.Messages[0].Data).to.exist
      const state = JSON.parse(viewStateResult.Messages[0].Data)
      expect(state.CrawlTasks).to.be.an('array')
      expect(state.CrawlTasks).to.have.lengthOf(0)
      expect(state.CrawlTasks[urls[0]]).to.not.exist
      expect(state.CrawlTasks[urls[1]]).to.not.exist
      expect(state.CrawlTasks[urls[2]]).to.not.exist
    })
  })
})
